lib/internal/per_context/primordials.js

```diff
@@ -1,34 +1,443 @@
+  } else {
+    copyProps(unsafe.prototype, safe.prototype)
+  }
+  copyProps(unsafe, safe)
+
+  ObjectSetPrototypeOf(safe.prototype, null)
+  ObjectFreeze(safe.prototype)
+  ObjectFreeze(safe)
+  return safe
+}
+primordials.makeSafe = makeSafe
+
+// Subclass the constructors because we need to use their prototype
+// methods later.
+// Defining the `constructor` is necessary here to avoid the default
+// constructor which uses the user-mutable `%ArrayIteratorPrototype%.next`.
+primordials.SafeMap = makeSafe(
+  Map,
+  class SafeMap extends Map {
+    constructor (i) { super(i) } // eslint-disable-line no-useless-constructor
+  }
+)
+primordials.SafeWeakMap = makeSafe(
+  WeakMap,
+  class SafeWeakMap extends WeakMap {
+    constructor (i) { super(i) } // eslint-disable-line no-useless-constructor
+  }
+)
+
+primordials.SafeSet = makeSafe(
+  Set,
+  class SafeSet extends Set {
+    constructor (i) { super(i) } // eslint-disable-line no-useless-constructor
+  }
+)
+primordials.SafeWeakSet = makeSafe(
+  WeakSet,
+  class SafeWeakSet extends WeakSet {
+    constructor (i) { super(i) } // eslint-disable-line no-useless-constructor
+  }
+)
+
+primordials.SafeFinalizationRegistry = makeSafe(
+  FinalizationRegistry,
+  class SafeFinalizationRegistry extends FinalizationRegistry {
+    // eslint-disable-next-line no-useless-constructor
+    constructor (cleanupCallback) { super(cleanupCallback) }
+  }
+)
+primordials.SafeWeakRef = makeSafe(
+  WeakRef,
+  class SafeWeakRef extends WeakRef {
+    // eslint-disable-next-line no-useless-constructor
+    constructor (target) { super(target) }
+  }
+)
+
+const SafePromise = makeSafe(
+  Promise,
+  class SafePromise extends Promise {
+    // eslint-disable-next-line no-useless-constructor
+    constructor (executor) { super(executor) }
+  }
+)
+
+primordials.PromisePrototypeCatch = (thisPromise, onRejected) =>
+  PromisePrototypeThen(thisPromise, undefined, onRejected)
+
+/**
+ * Attaches a callback that is invoked when the Promise is settled (fulfilled or
+ * rejected). The resolved value cannot be modified from the callback.
+ * Prefer using async functions when possible.
+ * @param {Promise<any>} thisPromise
+ * @param {() => void) | undefined | null} onFinally The callback to execute
+ *        when the Promise is settled (fulfilled or rejected).
+ * @returns {Promise} A Promise for the completion of the callback.
+ */
+primordials.SafePromisePrototypeFinally = (thisPromise, onFinally) =>
+  // Wrapping on a new Promise is necessary to not expose the SafePromise
+  // prototype to user-land.
+  new Promise((a, b) =>
+    new SafePromise((a, b) => PromisePrototypeThen(thisPromise, a, b))
+      .finally(onFinally)
+      .then(a, b)
+  )
+
+primordials.AsyncIteratorPrototype =
+  primordials.ReflectGetPrototypeOf(
+    primordials.ReflectGetPrototypeOf(
+      async function * () {}).prototype)
+
+ObjectSetPrototypeOf(primordials, null)
+ObjectFreeze(primordials)
+
+module.exports = primordials
```

Comment:
Do you think it's worth it porting all of primordials?

Reply:
Probably not, also some primordials won't be available on Node.js 14.x so we'd need to polyfill them. I've copied the file from Node.js source to be sure it wasn't the cause of the test failures, but that's not the final shape of this PR.